www/content/workstation/upgrade_labs.md

```diff
@@ -0,0 +1,455 @@
+
+Tracking and testing changes in a CI/CD pipeline is an important part of managing your Chef infrastructure. CI/CD pipelines and test-driven development (TDD) which are beyond the scope of this guide, see [Learn Chef Rally](https://learn.chef.io/) for tutorials and contact [Chef Software customer support](https://www.chef.io/support/) when you are ready to modernize your system.
+
+```
+Changes made to the following cookbooks in ./node-MYNODE-repo/cookbooks
+cannot be saved to the cookbook's source, though they can still be uploaded
+to a Chef Server:
+
+  - logrotate (v1.9.2)
+  - windows (v1.44.1)
+  - chef_handler (v1.4.0)
+
+You're ready to begin!
+
+Start with 'kitchen converge'.  As you identify issues, you
+can modify cookbooks in their original checkout locations or
+in the repository's cookbooks directory and they will be picked
+up on subsequent runs of 'kitchen converge'.
+```
+
+## Detect and Correct Cookbook Errors
+
+{{< note >}}
+Be cautious when running your cookbooks locally. It is possible to accidentally impact production infrastructure based on settings in your cookbooks and attributes. Read over your cookbooks and attributes, especially attributes set in roles and environments.  If needed, override attributes to be appropriate for local testing in your `kitchen.yml`.
+{{< /note >}}
+
+### Increment the Chef Infra Client Version
+
+In the `kitchen.yml` file, change the `product_version` line to `16`:
+
+```yaml
+provisioner:
+  name: chef_zero
+  product_name: chef
+  product_version: 16  # Change this line
+```
+
+If needed, you can "step forward" by first going from 12 to 13, correcting issues, then 13 to 14, etc.
+
+### Attempt a Converge and Check for Errors
+
+Save the file and test it in the `node-MY_NODE-repo` directory by running:
+
+```
+chef exec kitchen converge
+```
+
+Watch for Chef Infra errors. If any occur, fix them.
+
+### Test and Correct with Cookstyle
+
+To check for version upgrade issues, run:
+
+```
+$ chef exec cookstyle cookbooks/my_cookbook
+```
+
+Repeat this process for each cookbook for that node.
+
+#### Using Cookstyle Auto-correct
+
+Use Cookstyle's auto-correct feature to fix style errors by adding the `-a` (for auto-correct) flag:
+
+```
+$ chef exec cookstyle -a cookbooks/my_cookbook
+```
+
+Other issues may require manual intervention and editing.
+Repeat this process for each cookbook that the node consumes.
+
+
+### Copy Data Bags
+
+If data bags are used on your Chef Server, you will need to to download the data_bags directory `data_bags` directory in your repository.
+Note that this command does not support embedded keys in credentials files. If you use embedded keys, move the key to a key file.
+
+```
+cd node-node-01-repo
+chef exec knife download data_bags --chef-repo-path . --profile old-server --key my-old-key.pem
+```
+
+### Remove Server Searches
+
```

Comment:
We need to be much more up front with users about the danger of moving nodes this way. It's very common for users to populate configs using search data. Moving a node to a new server will wipe out those configs and has a high likelihood to cause production outages.

Reply:
I've added some suggested warning verbiage in a comment in the About this Guide section.